tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/MailboxProcessorType.fs

```diff
@@ -356,3 +365,41 @@ type MailboxProcessorType() =
         System.Threading.Thread.Sleep(5000) // cancellation after 500 pause for 5 seconds 
         if not gotGood || not gotBad then 
             failwith <| sprintf "Exected both good and bad async's to be cancelled afteMailbox should not fail!  gotGood: %A, gotBad: %A" gotGood gotBad
+
+    [<Fact>]
+    member this.StartImmediateStartsOnCurrentThread() =
+        /// Gets the current thread's ID and name
+        let getThreadInfo () =
+            let currentThread = Thread.CurrentThread
+            
+            { Id = currentThread.ManagedThreadId
+              Name = currentThread.Name }
+
+        // Although the ManagedThreadId should be unique, go ahead and set the calling thread
+        // name to something specific prior to starting the MailboxProcessor
+        Thread.CurrentThread.Name <- "This is the thread that starts the MailboxProcessor"
+
+        // Capture the ID and name of the calling thread
+        let callingThreadInfo = getThreadInfo ()
+
+        // Start a MailboxProcessor with StartImmediate and have it wait for a single message
+        // requesting the information of the thread that it is running on
+        let mailbox = MailboxProcessor<StartImmediateMessage>.StartImmediate(fun inbox -> async{
+            // Get the MailboxProcessor's thread immediately after starting it
+            let threadInfo = getThreadInfo ()
+            
+            // Block until a single message is received
+            let! message = inbox.Receive()
+
+            // Reply with the MailboxProcessor's thread information
+            match message with
+            | GetThreadInfo reply -> reply.Reply threadInfo
```

Comment:
It's been more than two years since I posted the original proposal, so it took me some time to recall what was it all about 😏. Thank you for taking over, btw!

My entire app logic was implemented by an async function used as an MP body. All the `main` function was doing is creating the MP and waiting for its body completion. I was thinking how to make the main thread useful by running the MP body rather than staying idle.

Given my limited knowledge of Async at that time, I assumed, that if I used `StartImmediate`

1. The main thread would be blocked until the body runs to completion.
2. The entire body would run on the same thread.

In the light of your SO answer, that the MP body only runs on the current thread until the first "real" async call, the second assumption isn't true, so it obviously won't cover your use case, where you need an entire MP body to be executed on the same thread. Also, it (almost) won't reduce the number of context switches, caused by continuation scheduling.

Since the number of use cases for the new MP method turned out to be somewhat limited, but you dedicated your time to finalizing and testing it, I guess it would be your call, whether to complete or abandon this feature.